### PR TITLE
Set adaptive=True for plotting of sympy.physics.continuum_mechanics.beam

### PR DESCRIPTION
**Pull request by Davide-sd**
#### References to other Issues or PRs

Another PR related to Issue #23036 

#### Brief description of what is fixed or changed

Enabled `adaptive=True` for plotting of loads of `sympy.physics.continuum_mechanics.beam`

#### Other comments

In #25469 the data series were changed to use a uniform meshing algorithm instead of the adaptive algorithm. Main motivations were performance and quality.

As explained in issue #25526 , the `sympy.physics.continuum_mechanics.beam` uses the `SingularityFunction` to create symbolic expression for the loads, which will be evaluated using SymPy instead of Numpy, making the evaluation slow. 

Until the improvement suggested in #25526 is implemented, for performance reasons the adaptive algorithm is better suited to deal with `SingularityFunction`.

#### Release Notes

* physics.continuum_mechanics
  * The adaptive algorithm will be used to plot the loads of a beam.

### CODE REVIEW

**Comment by moorepants** [comment]
Can you provide and example of what the new plots look like?

**Comment by Davide-sd** [comment]
> Can you provide and example of what the new plots look like?

Before #25469 the plots were generated with the adaptive algorithm. After #25469, they were generated with a uniform meshing algorithm, which is slower in this particular case. This PR brings them back to an adaptive algorithm. Hence, the plots look exactly like before!

If it's possible to solve issue #25526 by making `lambdify` to convert `SingularityFunction` into `Piecewise`, than this PR could be discarded.

**Comment by moorepants** [comment]
If #25469 is breaking plots, then shouldn't you be sure it isn't breaking plots before merging it?

**Comment by Davide-sd** [comment]
> If #25469 is breaking plots, then shouldn't you be sure it isn't breaking plots before merging it?

#25469 is not breaking anything. If you have **PROOF ABOUT BREAKING**, provide them!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! If you are unable or you don't want to, keep your allegations for yourself. I made a tremendous effort to implement a better plotting module while maintaining back-compatibility, which is measured by the amount of tests developed in each sympy's submodule. NO TEST FAILED, hence the PR was merged.

In short, that PR simply switches the algorithms for the discretization of the domain. Because of the reasons CLEARLY explained in issue #25526, the evaluation of `SingularityFunction` is extremely slow, which makes the adaptive algorithm better suited for the job.

You are free to express your opinion, but allegations NEEDS to be proven. Please, don't waste my time any longer if you are unwilling to understand the underlying problem that this PR attempts to solve.

**Comment by oscarbenjamin** [comment]
Let's all keep civil. I'm sure there is just misunderstanding.

@moorepants It is usually difficult (impossible?) to make deep changes that fix major things without also causing some problems elsewhere.

@Davide-sd I am struggling a little bit to understand from the OP what the impact of the various changes is. Let me try to describe what I think has happened from reading the above:

The PR gh-25469 moved the plotting module from using `experimental_lambdify` to `lambdify` which is something that everyone has always agreed would be a good idea because we don't want to maintain two separate `lambdify` functions at least not if `experimental_lambdify` is not going to be something that fixes any of the more major problems of `lambdify`. Ultimately what you have identified is that a basic improvement for plotting is to use `lambdify` with uniform meshing which is what gh-25469 changed it to by default. The uniform meshing is better partly because it is faster but also just because the current adaptive algorithm is not very good and so it mostly makes sense to use simpler, faster uniform meshing.

In this case what you have identified is that `SingularityFunction` is something that `lambdify` cannot currently handle but with `adaptive=False` as the new default the uniform meshing might be slower (is that right)? The timings that you show don't look like a huge slowdown (or speedup?): adaptive=True looks about 4x faster (which is obviously good but it's not like a 1000x massive performance regression or something)

In any case the sort of piecewise linear plots that you show in gh-25526 are a particular sort of unusual case where adaptive sampling (even using the current algorithm which is not great) happens to be usually better even though that is atypical for most plotting.

So basically it makes sense for the beam module to use `adaptive=True` even if that is not the best default for the plotting module in general. Is that correct?

**Comment by moorepants** [comment]
Dear David, I apologize if I offended you. It was not my intention. We have all put in an inordinate amount of time to improve and maintain SymPy. I have spent at least a a a couple hundred hours myself reviewing PRs and mentoring students to develop the continuum_mechanics package.

This PR and the related issue seemed to imply that the plots in beam.py no longer work as they did before and that someone else should fix it (by fixing lambdify). Thus, my question in the prior comment. I may have misunderstood, as I've not followed the plotting changes you are making and don't intend to.

I will not reply more than this, given your response. I wish you the best.

**Comment by Davide-sd** [comment]
> I will not reply more than this, given your response. I wish you the best.

Dear Jason, apology accepted. Now, it's my turn to apologize for the rough tone of my reply, in which I should have used a more diplomatic touch. The reason for the roughness was the false accusation of breaking something. Please, understand that if someone unaware of the refactoring of the plotting module comes into this thread and reads your message, it could trigger a "he said/she said" situation, which is useful to no one and waste my time for nothing, time that I would gladly spend developing stuffs rather than repeating myself over and over. I also understand that there is a language barrier between us: my English is far from perfect, which could make for a difficult communication. Apologies for that.

If you have observations or needs more clarifications, feel free to ask them: we indeed share the common goal of improving SymPy, which requires communication (I'm definitely unable to read people's mind).

Sincerely,

Davide.

**Comment by Davide-sd** [comment]
> The PR [gh-25469](https://github.com/sympy/sympy/pull/25469) moved the plotting module from using `experimental_lambdify` to `lambdify` which is something that everyone has always agreed would be a good idea because we don't want to maintain two separate `lambdify` functions at least not if `experimental_lambdify` is not going to be something that fixes any of the more major problems of `lambdify`. Ultimately what you have identified is that a basic improvement for plotting is to use `lambdify` with uniform meshing which is what [gh-25469](https://github.com/sympy/sympy/pull/25469) changed it to by default. The uniform meshing is better partly because it is faster but also just because the current adaptive algorithm is not very good and so it mostly makes sense to use simpler, faster uniform meshing.

Correct.

> In this case what you have identified is that `SingularityFunction` is something that `lambdify` cannot currently handle but with `adaptive=False` as the new default the uniform meshing might be slower (is that right)? The timings that you show don't look like a huge slowdown (or speedup?): adaptive=True looks about 4x faster (which is obviously good but it's not like a 1000x massive performance regression or something)

Indeed, for that particular case it's only 4x speedup/slowdown, which would be uninfluential for a single plot. However, [plot_loading_results](https://docs.sympy.org/latest/modules/physics/continuum_mechanics/beam.html#sympy.physics.continuum_mechanics.beam.Beam.plot_loading_results) creates 4 subplots: with `adaptive=False` it start to feel a little laggy.

> In any case the sort of piecewise linear plots that you show in [gh-25526](https://github.com/sympy/sympy/issues/25526) are a particular sort of unusual case where adaptive sampling (even using the current algorithm which is not great) happens to be usually better even though that is atypical for most plotting.

Not really. The adaptive algorithm mainly makes sense in this module because it is faster. From a quality standpoint, I believe the uniform meshing algorithm is still better, as pointed out in #25526 (the final example with the artifact). Because the `beam` module was already using the adaptive algorithm, I don't think that the quality improvements are worth a 4x slowdown. Hence, this PR which re-enable the adaptive algorithm for this module.

But, if it's easier to make `lambdify` converts `SingularityFunction` into a `Piecewise`, we should go for that and close this PR (without merging it). Unfortunately, I don't feel comfortable to play with `lambdify`-related printers.

> So basically it makes sense for the beam module to use `adaptive=True` even if that is not the best default for the plotting module in general. Is that correct?

Correct.

**Comment by oscarbenjamin** [comment]
> Unfortunately, I don't feel comfortable to play with `lambdify`-related printers.

Another possibility here is just to convert these functions in the plotting module before calling `lambdify`:
```python
In [5]: expr1 = 13750*SingularityFunction(x, 0, 0) - 5000*SingularityFunction(x, 2, 0) - 10000*Singul
   ...: arityFunction(x, 4, 1) + 31250*SingularityFunction(x, 8, 0) + 10000*SingularityFunction(x, 8,
   ...:  1)

In [6]: expr1
Out[6]: 
         0               0                1                0                1
13750⋅<x>  - 5000⋅<x - 2>  - 10000⋅<x - 4>  + 31250⋅<x - 8>  + 10000⋅<x - 8> 

In [7]: expr1.rewrite(SingularityFunction, Piecewise)
Out[7]: 
      ⎛⎧1  for x > 0⎞        ⎛⎧1  for x > 2⎞         ⎛⎧1  for x > 8⎞         ⎛⎧x - 8  for x > 8⎞   
13750⋅⎜⎨            ⎟ - 5000⋅⎜⎨            ⎟ + 31250⋅⎜⎨            ⎟ + 10000⋅⎜⎨                ⎟ - 
      ⎝⎩0  otherwise⎠        ⎝⎩0  otherwise⎠         ⎝⎩0  otherwise⎠         ⎝⎩  0    otherwise⎠   

      ⎛⎧x - 4  for x > 4⎞
10000⋅⎜⎨                ⎟
      ⎝⎩  0    otherwise⎠
```
Obviously though that only helps with plotting and not with any other use of `lambdify` that might involve singularity functions.

**Comment by oscarbenjamin** [comment]
If you did want to try changing `lambdify` then this is where the rewrites happen:
https://github.com/sympy/sympy/blob/691b1ab5f451c74e1761c0dbc23c89c20afda608/sympy/printing/codeprinter.py#L69-L78
Various things are already rewritten to Piecewise:
https://github.com/sympy/sympy/blob/691b1ab5f451c74e1761c0dbc23c89c20afda608/sympy/printing/codeprinter.py#L98-L100

**Comment by oscarbenjamin** [comment]
I've just checked and this looks fine.

@Davide-sd do you want to try improving `lambdify` or just merge this now?

At least for the Beam module this preserves the behaviour from before https://github.com/sympy/sympy/pull/25469 and it is good to preserve existing behaviour if we are not sure exactly that an alternative is better. On the other hand fixing `lambdify` might not be that difficult.

Obviously if this was merged and then if `lambdify` was subsequently improved to handle this then the decision about `adaptive=True` could be revisited.

**Comment by Davide-sd** [comment]
> I've just checked and this looks fine.
> 
> @Davide-sd do you want to try improving `lambdify` or just merge this now?

I'll give it a try to improve `lambdify` tomorrow. Let's keep this PR open for the moment... Thanks for pointing me in the right direction.

**Comment by sylee957** [comment]
I'm a bit misguided about the problem about `adaptive` though, 

For example, I had discovered some `bug` in adaptive before that makes it unreasonably inaccurate
and this can be worth experimenting

https://github.com/sympy/sympy/issues/19961

**Comment by smichr** [comment]
closing since SingularityFunction can now be re-written in terms of Piecewise per #25538
